src/cuda/poisson_fft.f90

```diff
@@ -0,0 +1,239 @@
+
+      allocate (poisson_fft%waves_dev(SZ, nx, (ny*(nz/2 + 1))/SZ))
+      poisson_fft%waves_dev = poisson_fft%waves
+
+      allocate (poisson_fft%ax_dev(nx), poisson_fft%bx_dev(nx))
+      allocate (poisson_fft%ay_dev(ny), poisson_fft%by_dev(ny))
+      allocate (poisson_fft%az_dev(nz), poisson_fft%bz_dev(nz))
+      poisson_fft%ax_dev = poisson_fft%ax; poisson_fft%bx_dev = poisson_fft%bx
+      poisson_fft%ay_dev = poisson_fft%ay; poisson_fft%by_dev = poisson_fft%by
+      poisson_fft%az_dev = poisson_fft%az; poisson_fft%bz_dev = poisson_fft%bz
+
+      allocate (poisson_fft%c_x_dev(nx*ny*(nz/2 + 1)))
+      allocate (poisson_fft%c_y_dev(nx*ny*(nz/2 + 1)))
+      allocate (poisson_fft%c_z_dev(nx*ny*(nz/2 + 1)))
+
+      ! We can't currently ask allocator to pass us an array with
+      ! exact shape we want, so we allocate an extra array here.
+      ! This will be removed when allocator is fixed.
+      allocate (poisson_fft%f_tmp(nz, SZ, nx*ny/SZ))
+
+      ! set cufft plans
+      ierrfft = cufftCreate(poisson_fft%planD2Zz)
+      ierrfft = cufftMakePlanMany(poisson_fft%planD2Zz, 1, nz, &
+                                  nz, 1, nz, nz/2+1, 1, nz/2+1, &
+                                  CUFFT_D2Z, nx*ny, worksize)
+      ierrfft = cufftSetWorkArea(poisson_fft%planD2Zz, poisson_fft%c_x_dev)
+
+      ierrfft = cufftCreate(poisson_fft%planZ2Dz)
+      ierrfft = cufftMakePlanMany(poisson_fft%planZ2Dz, 1, nz, &
+                                  nz/2+1, 1, nz/2+1, nz, 1, nz, &
+                                  CUFFT_Z2D, nx*ny, worksize)
+      ierrfft = cufftSetWorkArea(poisson_fft%planZ2Dz, poisson_fft%c_x_dev)
+
+      ierrfft = cufftCreate(poisson_fft%planZ2Zy)
+      ierrfft = cufftMakePlanMany(poisson_fft%planZ2Zy, 1, ny, &
+                                  ny, 1, ny, ny, 1, ny, &
+                                  CUFFT_Z2Z, nx*(nz/2 + 1), worksize)
+      ierrfft = cufftSetWorkArea(poisson_fft%planZ2Zy, poisson_fft%c_x_dev)
+
+      ierrfft = cufftCreate(poisson_fft%planZ2Zx)
+      ierrfft = cufftMakePlanMany(poisson_fft%planZ2Zx, 1, nx, &
+                                  nx, 1, nx, nx, 1, nx, &
+                                  CUFFT_Z2Z, ny*(nz/2 + 1), worksize)
+      ierrfft = cufftSetWorkArea(poisson_fft%planZ2Zx, poisson_fft%c_y_dev)
+
+   end function init
+
+   subroutine fft_forward_cuda(self, f)
+      implicit none
+
+      class(cuda_poisson_fft_t) :: self
+      class(field_t), intent(in) :: f
+
+      real(dp), device, pointer, dimension(:, :, :) :: f_dev
+      complex(dp), device, dimension(:, :, :), pointer :: c_x_ptr, c_y_ptr, &
+                                                          c_z_ptr
+
+      type(dim3) :: blocks, threads
+      integer :: ierrfft
+
+      select type(f); type is (cuda_field_t); f_dev => f%data_d; end select
+
+      ! First reorder f into cartesian-like data structure
+      blocks = dim3(self%nz/SZ, self%nx*self%ny/SZ, 1)
+      threads = dim3(SZ, SZ, 1)
```

Comment:
This pattern (or similar) occurs several times. Potentially worth capturing as a function/subroutine.

Reply:
I might even capture each commented "block" of code as a subroutine/function - the comment implies that there is a named concept, but this is personal preference

Reply:
Can be addressed as part of #73